File: gms-server/scripts-zh-CN/npc/gachapon.js
/*
	This file is part of the OdinMS Maple Story Server
    Copyright (C) 2008 Patrick Huy <patrick.huy@frz.cc>
		       Matthias Butz <matze@odinms.de>
		       Jan Christian Meyer <vimes@odinms.de>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as
    published by the Free Software Foundation version 3 as published by
    the Free Software Foundation. You may not use, modify or distribute
    this program under any other version of the GNU Affero General Public
    License.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
/* NPC Base
	Map Name (Map ID)
	Extra NPC info.
 */

var status;
var ticketId = 5220000;
var curMapName = "";

function start() {
    status = -1;
    // 用客户端的地图名宏，省掉一张与 NPC ID 强耦合的硬编码地名表
    curMapName = "#m" + cm.getMapId() + "#";

    action(1, 0, 0);
}

function action(mode, type, selection) {
    if (mode < 0) {
        cm.dispose();
    } else {
        if (mode == 1) {
            status++;
        } else {
            status--;
        }
        if (status == 0 && mode == 1) {
            if (cm.haveItem(ticketId)) {
                cm.sendSimple("你可以使用" + curMapName + "快乐百宝箱，请确保物品栏有足够的空位。\r\n\r\n#L1#抽一次#l\r\n#L10#抽#r十#k次#l");
            } else {
                cm.sendSimple("欢迎来到" + curMapName + "快乐百宝箱。我可以为您做些什么呢？\r\n\r\n#L0#什么是快乐百宝箱？#l\r\n#L1#在哪里可以购买快乐百宝券？#l");
            }
        } else if (status == 1 && cm.haveItem(ticketId)) {
            // 扣券与背包校验都在 Java 侧逐抽进行，这里不再预扣
            cm.doGachapon(selection == 10 ? 10 : 1, ticketId);
            cm.dispose();
        } else if (status == 1) {
            if (selection == 0) {
                cm.sendNext("玩转快乐百宝箱，赢得稀有卷轴、装备、椅子、熟练书和其他酷炫物品！你只需要一张 #b快乐百宝券#k 就有机会成为随机物品的幸运获得者。");
            } else {
                cm.sendNext("快乐百宝券可以在#r现金商店#k购买，可以使用NX或枫叶点购买。点击屏幕右下角的红色商店图标进入#r现金商店#k，就能购买快乐百宝券。");
            }
        } else if (status == 2) {
            cm.sendNextPrev("你会在" + curMapName + "的快乐百宝箱中找到各种物品，但最有可能找到与" + curMapName + "相关的物品和卷轴。");
        } else {
            cm.dispose();
        }
    }
}

File: gms-server/src/main/java/org/gms/server/gachapon/Gachapon.java
/*
 This file is part of the OdinMS Maple Story Server
 Copyright (C) 2008 Patrick Huy <patrick.huy@frz.cc>
 Matthias Butz <matze@odinms.de>
 Jan Christian Meyer <vimes@odinms.de>

 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU Affero General Public License as
 published by the Free Software Foundation version 3 as published by
 the Free Software Foundation. You may not use, modify or distribute
 this program under any other version of the GNU Affero General Public
 License.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU Affero General Public License for more details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.gms.server.gachapon;

import lombok.Getter;
import org.gms.client.Character;
import org.gms.constants.id.NpcId;
import org.gms.util.I18nUtil;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.gms.server.ItemInformationProvider;
import org.gms.util.Randomizer;

import java.util.Arrays;

/**
 * @author Alan (SharpAceX)
 */
public class Gachapon {
    private static final Logger log = LoggerFactory.getLogger(Gachapon.class);
    private static final Gachapon instance = new Gachapon();

    public static Gachapon getInstance() {
        return instance;
    }

    public enum GachaponType {

        GLOBAL(-1, -1, -1, -1, null, new Global()),
        HENESYS(NpcId.GACHAPON_HENESYS, 90, 8, 2, "GachaCommand.message2", new Henesys()),
        ELLINIA(NpcId.GACHAPON_ELLINIA, 90, 8, 2, "GachaCommand.message3", new Ellinia()),
        PERION(NpcId.GACHAPON_PERION, 90, 8, 2, "GachaCommand.message4", new Perion()),
        KERNING_CITY(NpcId.GACHAPON_KERNING, 90, 8, 2, "GachaCommand.message5", new KerningCity()),
        SLEEPYWOOD(NpcId.GACHAPON_SLEEPYWOOD, 90, 8, 2, "GachaCommand.message6", new Sleepywood()),
        MUSHROOM_SHRINE(NpcId.GACHAPON_MUSHROOM_SHRINE, 90, 8, 2, "GachaCommand.message7", new MushroomShrine()),
        SHOWA_SPA_MALE(NpcId.GACHAPON_SHOWA_MALE, 90, 8, 2, "GachaCommand.message8", new ShowaSpaMale()),
        SHOWA_SPA_FEMALE(NpcId.GACHAPON_SHOWA_FEMALE, 90, 8, 2, "GachaCommand.message9", new ShowaSpaFemale()),
        LUDIBRIUM(NpcId.GACHAPON_LUDIBRIUM, 90, 8, 2, "GachaCommand.message15", new Ludibrium()),
        NEW_LEAF_CITY(NpcId.GACHAPON_NLC, 90, 8, 2, "GachaCommand.message10", new NewLeafCity()),
        EL_NATH(NpcId.GACHAPON_EL_NATH, 90, 8, 2, "GachaCommand.message16", new ElNath()),
        LEAFRE(NpcId.GACHAPON_LEAFRE, 90, 8, 2, "GachaCommand.message17", new Leafre()),
        NAUTILUS_HARBOR(NpcId.GACHAPON_NAUTILUS, 90, 8, 2, "GachaCommand.message11", new NautilusHarbor());

        private static final GachaponType[] values = GachaponType.values();

        private final GachaponItems gachapon;
        @Getter
        private final int npcId;
        private final int common;
        private final int uncommon;
        private final int rare;
        /** i18n 键；GLOBAL 是兜底奖池、不对应任何扭蛋机，为 null */
        private final String nameKey;

        GachaponType(int npcid, int c, int u, int r, String nameKey, GachaponItems g) {
            this.npcId = npcid;
            this.gachapon = g;
            this.common = c;
            this.uncommon = u;
            this.rare = r;
            this.nameKey = nameKey;
        }

        private int getTier() {
            int chance = Randomizer.nextInt(common + uncommon + rare) + 1;
            if (chance > common + uncommon) {
                return 2; //Rare
            } else if (chance > common) {
                return 1; //Uncommon
            } else {
                return 0; //Common
            }
        }

        public int[] getItems(int tier) {
            return gachapon.getItems(tier);
        }

        public int getItem(int tier) {
            int[] gacha = getItems(tier);
            int[] global = GLOBAL.getItems(tier);
            int chance = Randomizer.nextInt(gacha.length + global.length);
            return chance < gacha.length ? gacha[chance] : global[chance - gacha.length];
        }

        public static GachaponType getByNpcId(int npcId) {
            for (GachaponType gacha : values) {
                if (npcId == gacha.npcId) {
                    return gacha;
                }
            }
            return null;
        }

        /** 真实存在的扭蛋机，按枚举顺序，不含兜底的 GLOBAL */
        private static GachaponType[] machines() {
            return Arrays.stream(values).filter(g -> g.nameKey != null).toArray(GachaponType[]::new);
        }

        public static String[] getLootNames() {
            return Arrays.stream(machines()).map(g -> I18nUtil.getMessage(g.nameKey)).toArray(String[]::new);
        }

        public static int[] getLootIds() {
            return Arrays.stream(machines()).mapToInt(GachaponType::getNpcId).toArray();
        }
    }

    public GachaponItem process(int npcId) {
        GachaponType gacha = GachaponType.getByNpcId(npcId);
        int tier = gacha.getTier();
        int item = gacha.getItem(tier);
        return new GachaponItem(tier, item);
    }

    public static class GachaponItem {
        private final int id;
        private final int tier;

        public GachaponItem(int t, int i) {
            id = i;
            tier = t;
        }

        public int getTier() {
            return tier;
        }

        public int getId() {
            return id;
        }
    }

    public static void log(Character player, int itemId, String map) {
        String itemName = ItemInformationProvider.getInstance().getName(itemId);
        log.info(I18nUtil.getLogMessage("Gachapon.log.info"), player.getName(), itemName, itemId, map);
    }
}
